### PR TITLE
Generate a simple non-bootable firmware when not provided.

### PR DESCRIPTION
**Pull request by Swordfish90**
Hello guys,

I'd really like to help with the HLE bios task #367. It seems the drastic bios file linked in the issue are working very well, but sadly the games won't work without a proper firmware.

With this pull request, I'm trying to salvage some work done in #527, generating a dummy firmware when none are provided.

Using the free drastic bios files (linked in #367) I've been able to play every game I tried, so it sounds like a promising start.

Let me know if you want me to change anything.

### CODE REVIEW

**Comment by RSDuck** [comment]
can you make the firmware configurable via a GUI dialog?

**Comment by Swordfish90** [comment]
@RSDuck What do you mean? Exposing settings like language and username?

**Comment by RSDuck** [comment]
exactly

**Comment by Swordfish90** [comment]
@RSDuck Added the two options into settings. Let me know if I missed something.

File: src/SPI.cpp [hunk 1]
```diff
@@ -162,6 +184,17 @@ void Reset()
 
     UserSettings = userdata;
 
+    // setting up username
+    int usernameSize = strlen(Config::FirmwareUserName);
+    Firmware[UserSettings+0x1A] = usernameSize;
+    for (int i = 0; i < usernameSize; ++i) {
+        Firmware[UserSettings + 0x06 + i * 2] = Config::FirmwareUserName[i];
```
Comment (RSDuck) [inline src/SPI.cpp:191]: this can't handle codepoints above first 256

File: src/frontend/qt_sdl/EmuSettingsDialog.cpp [hunk 2]
```diff
@@ -252,6 +253,35 @@ void EmuSettingsDialog::on_btnFirmwareBrowse_clicked()
...
+    QLineEdit* nameEdit = new QLineEdit(&dialog);
+    nameEdit->setText(Config::FirmwareUserName);
+    form.addRow("Username", nameEdit);
+
+    QStringList languages { "Auto", "Japanese", "English", "French", "German", "Italian", "Spanish" };
+
+    QComboBox* languagesBox = new QComboBox(&dialog);
+    languagesBox->addItems(languages);
+    languagesBox->setCurrentIndex(Config::FirmwareLanguage + 1);
+    form.addRow("Language", languagesBox);
```
Comment (RSDuck) [inline src/frontend/qt_sdl/EmuSettingsDialog.cpp:270]: could you add the remaining settings (birthday, message and favourite color) as well?

File: src/SPI.cpp [hunk 3]
```diff
@@ -162,6 +184,17 @@ void Reset()
 
     UserSettings = userdata;
 
+    // setting up username
```
Comment (RSDuck) [inline src/SPI.cpp:187]: please make this optional if a firmware file is present

**Review by RSDuck — CHANGES_REQUESTED** [review]
needs a few additions, otherwise it's good!

**Comment by Swordfish90** [comment]
@RSDuck I just pushed the requested changes. Now the settings are applied only when loading the dummy firmware or when the user checks "Override firmware settings" which default to false.

**Comment by RSDuck** [comment]
one last thing I should have asked earlier: can you implement the dialog via a QT xml file and structure it like the other dialogs? 

https://github.com/Arisotura/melonDS/blob/master/src/frontend/qt_sdl/main.cpp#L1775
https://github.com/Arisotura/melonDS/blob/master/src/frontend/qt_sdl/VideoSettingsDialog.h#L37

**Comment by Swordfish90** [comment]
No problem. I'll do that right away!

**Comment by kyandora** [comment]
The username can overflow if it's larger than 10 characters (https://github.com/Swordfish90/melonDS/pull/3).
Although the strings are std::u16string katakana won't show up properly.

**Comment by Swordfish90** [comment]
@RSDuck Done. Feel free to let me know if something else is missing.

File: src/frontend/qt_sdl/FirmwareSettingsDialog.h [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+namespace Ui { class FirmwareSettingsDialog; }
+class FirmwareSettingsDialog;
+
+class FirmwareSettingsDialog : public QDialog
+{
+    Q_OBJECT
+
+public:
+    const QStringList colours {
```
Comment (RSDuck) [inline src/frontend/qt_sdl/FirmwareSettingsDialog.h:33]: that curly brace belongs to the next line

File: src/frontend/qt_sdl/FirmwareSettingsDialog.h [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        "Turqoise",
+        "Light Blue",
+        "Blue",
+        "Dark Blue",
+        "Dark Purple",
+        "Light Purple",
+        "Dark Pink"
+    };
+
+    const QStringList languages {
```
Comment (RSDuck) [inline src/frontend/qt_sdl/FirmwareSettingsDialog.h:52]: same here

**Comment by kyandora** [comment]
Would configuration pollution be an issue? I've got a patch for the strings, but mangling it config.cpp compliant requires 36 integers.

**Comment by RSDuck** [comment]
can you elaborate?

**Comment by kyandora** [comment]
FirmwareUsername and FirmwareMessage are stored as unsigned shorts. During configuration QT parses and saves the strings in UTF-16. The result can be memcopied into the firmware.
As melonDS stores it's settings in chars and integers, the arrays would need to be split into single shorts (`{"FirmwareMessage00", 0, &FirmwareMessage[0], 68, NULL, 0},`).

**Comment by RSDuck** [comment]
I see, though I think it's preferable to save the strings as UTF-8 because it fits better with the rest of the code and to keep the configs editable by hand

**Comment by RSDuck** [comment]
@Swordfish90 and @kyandora can we merge this?

**Comment by Swordfish90** [comment]
@RSDuck I think so. Yesterday I applied the changes to the curly braces (my religion is different :smile:), so if you like the dialog layout we're good to go.

**Comment by kyandora** [comment]
Thanks to our previously unused iconv dependency, the strings are now proper UTF-16LE. https://github.com/Swordfish90/melonDS/pull/4

**Comment by YoshiRulz** [comment]
Something like this would significantly reduce the work required to make an EmuHawk integration MVP.

**Comment by fancythedeveloper** [comment]
MVP?

**Comment by YoshiRulz** [comment]
Minimum viable product. We just need roms, firmware, video, input, and savestates.

**Comment by rafaelvcaetano** [comment]
I don't know if there are still plans to merge this or not, but I've been working on a similar implementation for the standalone Android version and whenever a ROM requires wireless communications, it just hangs. This causes some games to be unplayable, like the Gen 4 Pokémon games because they perform some wireless related operations in the main menu screen, making it impossible to enter the game. I've gotten around it by modifying some values in the firmware's Wifi Calibration/Settings section. Local multiplayer seems to retain its functionality but online connectivity seems be broken when configuring an access point. Haven't tested these changes in the PC version yet.

**Comment by RSDuck** [comment]
sorry for lettings this wait for so long, but I think the UTF8 -> UTF16 conversion is still done incorrectly. I suggest using https://en.cppreference.com/w/cpp/locale/codecvt_utf8_utf16 for it (eventhough it's deprecated, but there's no good replacement in the std library currently, so might as well use it).

**Comment by vadosnaprimer** [comment]
Did https://github.com/Arisotura/melonDS/pull/1174 resolve this PR?

**Comment by RayyanAnsari** [comment]
no the updated version of this PR is https://github.com/Arisotura/melonDS/pull/1175

**Comment by Arisotura** [comment]
well this seems to be no longer needed

**Pull request closed without merging**